doc: add allocation page to book (#652)
tachyon-zcash/ragu: book/src/guide/primitives/allocation.md

# Allocation

All wires are created in groups whenever new gates are created via
[`mul()`] (or, more generally, with [`gate()`]). In some rare cases,
gadgets need _standalone_ wires that cannot be easily represented as

> linear combinations of existing wires. The [`Allocator`] trait

Reviewer: Ah, even better and more accurate. Perfection is key!

```suggestion
gadgets need _standalone_ wires that cannot be represented as linear
combinations of wires they have access to. The [`Allocator`] trait
```

# Allocation

All wires are created in groups whenever new gates are created via
[`mul()`] (or, more generally, with [`gate()`]). In some rare cases,
gadgets need _standalone_ wires that cannot be easily represented as
linear combinations of existing wires. The [`Allocator`] trait
generalizes over (possibly stateful) strategies for obtaining free

> wires to witness arbitrary data.

Reviewer: ```suggestion
generalizes over (possibly stateful) strategies for assigning
arbitrary data to free wires.
```

Reviewer: I can't figure out why I prefer this wording. I think it has something to do with the subject-object distinction; we _witness_ data _by_ assigning _to_ wires -- or something like that.

## The [`Allocator`] Trait {#allocator-trait}

```rust,ignore
pub trait Allocator<'dr, D: Driver<'dr>> {
    fn alloc(
        &mut self,
        dr: &mut D,
        value: impl Fn() -> Result<Coeff<D::F>>,
    ) -> Result<D::Wire>;

>     fn donate(&mut self, _extra: D::Extra) {}

Reviewer: I support omitting explanation of this until later in the article, though it doesn't matter that much. I think it's always okay to lie about the full trait description in the book for the purposes of exposition.

}
```

`Allocator::alloc` takes a mutable reference to the driver and a
witness-producing closure, and returns a single wire. The closure
follows the same purity contract as [`Driver::mul`]: it may be called

> zero or more times and must be side-effect-free. Where `Driver::mul`

Reviewer: I don't like starting sentences with "Where"

produces three wires from a multiplication relationship, `alloc`
produces one wire carrying an arbitrary value.

The `donate` method accepts a spare
[`Extra`](ragu_core::drivers::DriverTypes::Extra) token from an
external gate whose $D$ wire is unconstrained. By default it drops
the token; advanced allocators override it to pool spare wires for

> reuse.

Reviewer: If we omit `donate` from the earlier trait description, it can be mentioned later when the `Extra` token actually has use / importance to the reader. It comes out of nowhere here.

Gadgets that need standalone wires accept a generic `&mut A` where
`A: Allocator`. For example, [`Element::alloc`] delegates to

> whichever allocator the caller provides:

Reviewer: We should word this: gadgets that need standalone wires should _prefer_ to accept a generic allocator. It is not always necessary if threading an allocator through the callsite is difficult. (This is subject to change; we may introduce a global allocator that the driver has possession of to simplify these patterns. In any case, the book should document what is currently implemented.)

> ```rust,ignore

Reviewer: What does `ignore` do in the book? It's needed in doc tests often because those are compiled by default, but the book shouldn't have any CI infra for that yet.

let x = Element::alloc(dr, &mut allocator, witness)?;
```

This lets the same gadget code work with any allocation strategy

> the caller chooses.

Reviewer: more accurately: This allows the caller to determine the optimal allocation strategy for gadgets that do not otherwise care or lack the context to make the best decision themselves.

## The Unit Allocator {#unit-allocator}

The simplest allocator is `()`. Each call invokes [`mul()`] with
$A = C = 0$ and returns the $B$ wire:

```rust,ignore
impl<'dr, D: Driver<'dr>> Allocator<'dr, D> for () {
    fn alloc(
        &mut self,
        dr: &mut D,
        value: impl Fn() -> Result<Coeff<D::F>>,
    ) -> Result<D::Wire> {
        let (_, b, _) = dr.mul(
            || Ok((Coeff::Zero, value()?, Coeff::Zero)),
        )?;
        Ok(b)
    }
}
```

Call site:

```rust,ignore
let x = Element::alloc(dr, &mut (), witness)?;

> ```

Reviewer: The call site varies from one allocator to another in such a trivial, mechanical way that these particular examples pollute the article. Remove.

One gate per wire, stateless, but wasteful when allocations come in

> pairs.

Reviewer: overly casual, partial sentence structure. Be precise: "In exchange for being stateless, this allocator creates a gate for every free wire, which is not optimal."

## [`SimpleAllocator`] {#simple-allocator}

The unit allocator wastes capacity because each gate allocates four
wire slots, but [`gate()`] returns three wires plus a spare `Extra`
token representing the $D$ wire. The gate is subject to
$A \cdot B = C$ and $C \cdot D = 0$. When $A = C = 0$, both
constraints are trivially satisfied and both $B$ and $D$ are

> unconstrained. A single gate therefore contains **two free slots**.

Reviewer: Let the previous section explain the downside of the unit allocator. Defeat it with this paragraph:

this is poor wording, but could inspire a better description:

> Theoretically, an allocator could offer two wires per gate by distributing the `a` and `b` wires of each `mul`, allowing `c` to be a (meaningless) product that is discarded. This is slightly suboptimal, because the trace contains the useless discarded product and is more expensive to manipulate. Instead, the more general [`gate()`] (which creates four wires) can be used, where the `b` and `d` wires are distributed and the `a` and `c` wires are assigned to zero.
>
> [`SimpleAllocator`] exploits this by creating gates, assigning the `b` wire, and then stashing the `d` wire's token for future allocation with the [driver's assistance.]

driver's assistance can link to the `DriverTypes` documentation which describes the `alloc_extra` function (or at least, it _should_ describe it)

[`SimpleAllocator`] exploits this by pairing consecutive allocations
into one gate:

1. **First call:** allocates a gate with $A = C = 0$, returns $B$,
   and stashes the `Extra` token for $D$.
2. **Second call:** redeems the stashed token via
   [`assign_extra`][assign-extra], returning $D$ without allocating
   a new gate.

```rust,ignore
impl<'dr, D: Driver<'dr>> Allocator<'dr, D>
    for SimpleAllocator<D::Extra>
{
    fn alloc(
        &mut self,
        dr: &mut D,
        value: impl Fn() -> Result<Coeff<D::F>>,
    ) -> Result<D::Wire> {
        if let Some(extra) = self.stash.take() {
            dr.assign_extra(extra, value)
        } else {
            let (_, b, _, extra) = dr.gate(
                || Ok((Coeff::Zero, value()?, Coeff::Zero)),
            )?;
            self.stash = Some(extra);
            Ok(b)
        }
    }
}

> ```

Reviewer: From the user's perspective, all of this code is useless. What it does is useful, to some extent, and the trade-off in terms of performance is relevant. But these details are not. Just say (roughly) "in exchange for being stateful, this is less wasteful."

Typical usage:

```rust,ignore
let allocator = &mut SimpleAllocator::new();
let a = Element::alloc(dr, allocator, witness_a)?;
let b = Element::alloc(dr, allocator, witness_b)?;
// Both a and b share one gate.

> ```

Reviewer: This is... somewhat okay as an illustration, so my earlier comment about the trivial example might not apply here.

```admonish info
Dropping a `SimpleAllocator` that still holds a stashed token is
safe. The driver already assigned $D = 0$ for that gate, so no
constraint is violated.

> ```

Reviewer: Let's omit this entirely.

## [`PoolAllocator`] {#pool-allocator}

Some gadgets allocate a gate whose $D$ wire is unconstrained because

> $C = 0$. Rather than waste that wire, they can _donate_ the `Extra`

Reviewer: More specifically c is _constrained_ to be zero somehow. Probably via `enforce_zero` but possibly by other methods.

token to the allocator. [`PoolAllocator`] collects these donated
tokens and hands them out on future `alloc` calls before falling
back to new gates.

> [`Boolean::alloc`] is the canonical example. The boolean constraint

Reviewer: LLMs hand out canonical too easily, I like to use it for specific situations like when we're talking about ambiguous representations of ideas or concepts. ("What is canon?") here it's just the only example we have _yet_.

Just say, "As an example, ..."

$a \cdot (1 - a) = 0$ produces a gate where $C = 0$, leaving $D$
unconstrained. `Boolean::alloc` donates that spare `Extra` token
back to the allocator:

```rust,ignore
pub fn alloc<A: Allocator<'dr, D>>(
    dr: &mut D,
    allocator: &mut A,
    value: DriverValue<D, bool>,
) -> Result<Boolean<'dr, D>> {
    // ... constrains a * (1 - a) = 0 ...

    // The gate's spare D wire is donated to the allocator.
    allocator.donate(extra);

    Ok(Boolean { value, wire: a })
}

> ```

Reviewer: hmm, this is 80% signature and 20% useful code.

```suggestion
```rust,ignore
    let (a, b, c, d) = dr.gate(|| /* ... */)?;
    dr.enforce_zero(|lc| lc.add(&c)); // c = 0

    // The gate's spare D wire is donated to the allocator.
    allocator.donate(d);
```
```

When using `PoolAllocator`, subsequent allocations redeem these
donated tokens at zero gate cost:

```rust,ignore
let allocator = &mut PoolAllocator::new();
let flag = Boolean::alloc(dr, allocator, bit)?;   // 1 gate
let x = Element::alloc(dr, allocator, witness)?;  // 0 gates (reuses donated D)
```

Like `SimpleAllocator`, `PoolAllocator` also pairs its own gate
allocations internally. When the pool is empty and a fresh gate is
needed, the spare `Extra` from that gate enters the pool for the
next call.

```admonish info
Dropping a `PoolAllocator` with tokens still in the pool is safe.
The driver already assigned $D = 0$ for those gates.

> ```

Reviewer: I think all of this can be removed, I would hope the reader has gotten the point by now.

> ## Choosing an Allocator {#choosing}

Reviewer: I'm not sure if the recommendations here are complete. There might be different trade-offs. Ultimately the choice of allocator is both a decision a gadget chooses (if they ask to receive one from the caller) and one the circuit developer chooses (if they're asked to provide one) and in each case the decision could be complicated.
